### PR TITLE
Raise ClaimInInvalidFormat properly

### PR DESCRIPTION
**Pull request by orenbm**
#### What does this PR do?
1. Raises a ClaimInInvalidFormat when the claim is actually invalid instead of only when the claim was valid. 
2. Parse the xms_mirid separately so it's clearer where the error came from.

### CODE REVIEW

File: app/domain/authentication/authn_azure/validate_application_identity.rb [hunk 1]
```diff
@@ -30,11 +31,9 @@ def call
 
       private
 
-      def validate_xms_mirid_format
+      def parse_xms_mirid
         begin
-          required_keys = %w(subscriptions resourcegroups providers)
-          missing_keys = required_keys - xms_mirid_hash.keys
-          raise Err::ClaimInInvalidFormat, "Required keys #{missing_keys} are missing" if missing_keys.empty?
+          xms_mirid_hash
```
Comment (eladkug) [inline app/domain/authentication/authn_azure/validate_application_identity.rb]: In this function(xms_mirid_hash)  you dont raise exception . You prefer to handle it separately on the caller function? The problem is we will not know that the problem was with Hash creating.

File: app/domain/authentication/authn_azure/validate_application_identity.rb [hunk 2]
```diff
@@ -30,14 +31,8 @@ def call
...
 
-      def validate_xms_mirid_format
-        begin
-          required_keys = %w(subscriptions resourcegroups providers)
-          missing_keys = required_keys - xms_mirid_hash.keys
-          raise Err::ClaimInInvalidFormat, "Required keys #{missing_keys} are missing" if missing_keys.empty?
-        rescue => e
-          raise Err::ClaimInInvalidFormat, e.inspect
-        end
+      def parse_xms_mirid
```
Comment (eladkug) [inline app/domain/authentication/authn_azure/validate_application_identity.rb]: why we need three function for it:
1. parse_xms_mirid call to 
2. xms_mirid_hash which call to
3. parsed_xms_mirid

i think that we can remove at least one function here

File: app/domain/errors.rb [hunk 3]
```diff
@@ -292,16 +292,21 @@ module AuthnAzure
           code: "CONJ00050E"
         )
 
-        ClaimInInvalidFormat = ::Util::TrackableErrorClass.new(
-          msg:  "xms_mirid claim has been received in an invalid format. Reason: {0}",
+        TokenFieldNotFoundOrEmpty = ::Util::TrackableErrorClass.new(
+          msg:  "Field '{0-field-name}' not found or empty in token",
```
Comment (eladkug) [inline app/domain/errors.rb]: not found or empty token

Comment (orenbm) [reply app/domain/errors.rb]: what do you mean?

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 2 above)
Comment (orenbm) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: Each one has its own purpose. The Command Class's `call` method needs to tell the story of what is going on - thus it has a call to the `parse_xms_mirid`.

`xms_mirid_hash` needs to be memoized so we don't do the work of `parsed_xms_mirid` every time. that's why it calls the `parsed_xms_mirid` method that does all the work and memoizes its output. 

the `parsed_xms_mirid` method does the actual parsing and has that name as in ruby we give methods that return an object the name of the returned object.

File: app/domain/errors.rb (hunk 3 above)
Comment (eladkug) [reply app/domain/errors.rb]: need to remove the word "in" (empty **in** token)

Comment (orenbm) [reply app/domain/errors.rb]: but it should be there :)

the field is missing inside the token. the token itself is not missing.

File: app/domain/authentication/authn_azure/validate_application_identity.rb [hunk 4]
```diff
@@ -49,23 +44,35 @@ def xms_mirid_hash
...
-            xms_mirid_hash["providers"] = split_xms_mirid[index + 1, index + 3]
+          index = 0
+          split_xms_mirid.each_with_object({}) do |property, xms_mirid_hash|
+            case property
+            when "subscriptions"
+              xms_mirid_hash["subscriptions"] = split_xms_mirid[index + 1]
+            when "resourcegroups"
+              xms_mirid_hash["resourcegroups"] = split_xms_mirid[index + 1]
+            when "providers"
+              xms_mirid_hash["providers"] = split_xms_mirid[index + 1, index + 3]
```
Comment (sgnn7) [inline app/domain/authentication/authn_azure/validate_application_identity.rb]: Drive-by comment: is a default case needed here?

File: app/domain/authentication/authn_azure/validate_application_identity.rb [hunk 5]
```diff
@@ -49,23 +44,35 @@ def xms_mirid_hash
...
+            end
+            index += 1
           end
-          index += 1
+        rescue => e
+          raise Err::XmsMiridParseError.new(@xms_mirid_token_field, e.inspect)
+        end
+      end
+
+      def validate_xms_mirid_format
```
Comment (orenbm) [inline app/domain/authentication/authn_azure/validate_application_identity.rb]: @sigalsax do you think that we should validate also that the value of `providers` has 3 parts?

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 4 above)
Comment (orenbm) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: Thanks for taking a look @sgnn7 ! There is actually no need for a default case as any other properties are not needed in the hash. We added this so we don't assume too much on the structure of the `xms_mirid` field and we assume only what we have to assume.

However, I do see a bug here as is needs to be `split_xms_mirid[index + 1, 3]`. The current code would have taken `index + 3` parts of `split_xms_mirid` and put them in the value for `xms_mirid_hash["providers"]`, as can be seen here - https://apidock.com/ruby/v1_9_3_125/Array/slice

@sigalsax  ^^

Comment (sgnn7) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: @orenbm Fair enough - I just wanted to make sure that we wouldn't have exceptions when we got some unexpected case values :)

File: app/domain/authentication/authn_azure/validate_application_identity.rb [hunk 6]
```diff
@@ -49,23 +44,35 @@ def xms_mirid_hash
...
+        required_keys = %w(subscriptions resourcegroups providers)
+        missing_keys = required_keys - xms_mirid_hash.keys
+        unless missing_keys.empty?
+          raise Err::MissingRequiredKeysInXmsMirid.new(missing_keys, @xms_mirid_token_field)
```
Comment (sigalsax) [inline app/domain/authentication/authn_azure/validate_application_identity.rb]: `MissingRequiredKeysInXmsMirid` please add this to the log issue for TWs
https://github.com/cyberark/conjur-docs/issues/742
Create a new section "Added logs" to the end so Shuli knows which new ones to review

File: app/domain/errors.rb [hunk 7]
```diff
@@ -292,16 +292,21 @@ module AuthnAzure
           code: "CONJ00050E"
         )
 
-        ClaimInInvalidFormat = ::Util::TrackableErrorClass.new(
```
Comment (sigalsax) [inline app/domain/errors.rb]: Please update here cyberark/conjur-docs#742 that you are removing this log

File: app/domain/errors.rb [hunk 8]
```diff
@@ -292,16 +292,21 @@ module AuthnAzure
...
-          msg:  "xms_mirid claim has been received in an invalid format. Reason: {0}",
+        TokenFieldNotFoundOrEmpty = ::Util::TrackableErrorClass.new(
+          msg:  "Field '{0-field-name}' not found or empty in token",
           code: "CONJ00051E"
         )
 
-        TokenFieldNotFoundOrEmpty = ::Util::TrackableErrorClass.new(
-          msg:  "Field '{0-field-name}' not found or empty in token",
+        XmsMiridParseError = ::Util::TrackableErrorClass.new(
```
Comment (sigalsax) [inline app/domain/errors.rb]: `XmsMiridParseError ` please add this to the log issue for TWs
cyberark/conjur-docs#742
Create a new section "Added logs" to the end so Shuli knows which new ones to review

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 4 above)
Comment (sigalsax) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: good catch!

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 5 above)
Comment (sigalsax) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: I don't think we should limit ourselves to 3 parts... regardless, we specify `index + 1, 3` so we will take the 3 elements and otherwise an exception will be caught in try/catch clause above

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 2 above)
Comment (sigalsax) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: Although the refactor is cleaner, was there a problem with the logic that this was refactored?

File: app/domain/authentication/authn_azure/validate_application_identity.rb [hunk 9]
```diff
@@ -30,14 +31,8 @@ def call
 
       private
 
-      def validate_xms_mirid_format
-        begin
-          required_keys = %w(subscriptions resourcegroups providers)
-          missing_keys = required_keys - xms_mirid_hash.keys
-          raise Err::ClaimInInvalidFormat, "Required keys #{missing_keys} are missing" if missing_keys.empty?
-        rescue => e
-          raise Err::ClaimInInvalidFormat, e.inspect
```
Comment (sigalsax) [inline app/domain/authentication/authn_azure/validate_application_identity.rb]: why was this removed in refactor? Don't we want to catch any errors that are raised?

File: app/domain/errors.rb [hunk 10]
```diff
@@ -292,16 +292,21 @@ module AuthnAzure
...
+          msg:  "Failed to parse xms_mirid {0}. Reason: {1}",
           code: "CONJ00052E"
         )
 
+        MissingRequiredKeysInXmsMirid = ::Util::TrackableErrorClass.new(
+          msg:  "Required keys {0} are missing in xms_mirid {1}",
```
Comment (sigalsax) [inline app/domain/errors.rb]: Why are we naming this "keys"? Maybe Fields? Properties?

Comment (sigalsax) [reply app/domain/errors.rb]: `MissingRequiredKeysInXmsMirid ` please add this to the log issue for TWs
cyberark/conjur-docs#742
Create a new section "Added logs" to the end so Shuli knows which new ones to review

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 2 above)
Comment (orenbm) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: Yes. we raised an error and rescued it straight after that so the original error was never raised

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 5 above)
Comment (orenbm) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: That's not entirely correct. In case of the following xms_mirid:
```
/subscriptions/<subscription-id>/resourcegroups/<some-value>/providers/<just_one_value>
```
then `split_xms_mirid[index + 1, 3]` will be `[<just_one_value>]`. So if the providers part is less than 3 parts then `split_xms_mirid[index + 1, 3]` will return less than 3 parts and the `providers` field will be invalid.

Comment (orenbm) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: More importantly - Why don't you think we should limit ourselves to 3 parts? We already do it in the code.

File: app/domain/errors.rb (hunk 8 above)
Comment (orenbm) [reply app/domain/errors.rb]: I added these 2 as a PR comment

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 9 above)
Comment (orenbm) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: I rescue errors in the new method for the parse. This way it's clearer which logic raised the error.

File: app/domain/errors.rb (hunk 10 above)
Comment (orenbm) [reply app/domain/errors.rb]: Changed to "fields"

File: app/domain/authentication/authn_azure/validate_application_identity.rb (hunk 5 above)
Comment (sigalsax) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: Okay, I was under the assumption that `split_xms_mirid[index + 1, 3]` but throw an index out of bounds exception.

Comment (sigalsax) [reply app/domain/authentication/authn_azure/validate_application_identity.rb]: So let's check for the length as well so that if Azure changes in the future, the user will know where the failure occured

File: app/domain/errors.rb [hunk 11]
```diff
@@ -292,16 +292,26 @@ module AuthnAzure
...
         )
 
+        MissingRequiredFieldsInXmsMirid = ::Util::TrackableErrorClass.new(
+          msg:  "Required fields {0} are missing in xms_mirid {1}",
+          code: "CONJ00053E"
+        )
+
+        MissingProviderFieldsInXmsMirid = ::Util::TrackableErrorClass.new(
+          msg:  "Provider fields are missing in xms_mirid {1}",
+          code: "CONJ00053E"
```
Comment (sigalsax) [inline app/domain/errors.rb]: update the error code to `CONJ00053E `

Comment (sigalsax) [reply app/domain/errors.rb]: Do we want to print the providers field that were extracted instead of the whole xms_mirid string?

Comment (orenbm) [reply app/domain/errors.rb]: No, I prefer that the user will get the whole field

**Comment by orenbm** [comment]
Closing this due to conflicts that raised when we rebased the `authn-azure` branch onto master. Created PR https://github.com/cyberark/conjur/pull/1384 instead.
